### PR TITLE
Pull in Tags 3.0; Create Pages 3.1 FAT

### LINKED ISSUE #21427: Pull in Jakarta Tags 3.0 for EE10

### PR DESCRIPTION
**Pull request by volosied**
fixes #21427 

Updated the JSP/Pages Implementation to handle new Tag Libraries. The API already exists under io.openliberty.jakarta/tags.3.0.bnd. This PR 
- created the implementation bundle as well as other internal builds. 
- Renamed the new bundles to use tags over jstl.
- Updated com.ibm.ws.jsp implementation to process the new jakarta.tag.* URIs and transformed other jsp bundles using ee10 rules. 
- Enabled EE10 Repeat over the com.ibm.ws.jsp.2.3_fat bucket. - Skipped several EL tests due to changes in the Pages 3.1 API.  Issue to re-enabled them  ->  https://github.com/OpenLiberty/open-liberty/issues/21743
- Created the new Pages 3.1 FAT Bucket  (io.openliberty.pages.3.1.internal_fat)

### CODE REVIEW

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

File: dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSP23JSP22ServerTest.java [hunk 1]
```diff
@@ -124,7 +124,7 @@ public void testJsp23to22FeatureChange() throws Exception {
      * @throws Exception
      */
     @Test
-    @SkipForRepeat(SkipForRepeat.NO_MODIFICATION)
+    @SkipForRepeat({SkipForRepeat.NO_MODIFICATION, SkipForRepeat.EE10_FEATURES})
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSP23JSP22ServerTest.java:127]: @volosied can you open an issue so we add an EE10 pages-3.1 test for this functionality? If you can provide a link here that would be good so I can ensure we track it. Thanks!

Comment (volosied) [reply dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSP23JSP22ServerTest.java:127]: This test should already be added under the new 3.1 FAT.

Comment (volosied) [reply dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSP23JSP22ServerTest.java:127]: Oops. It's not in this PR because I never pulled it in from my other branch. I'll add it in soon.

Comment (volosied) [reply dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSP23JSP22ServerTest.java:127]: Test added

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Comment by tevans78** [comment]
@volosied please remember to squash/fixup your commits before final delivery of this PR

File: dev/com.ibm.ws.jsp.2.3/transformed10.bnd [hunk 2]
```diff
@@ -8,11 +8,11 @@
...
 #     IBM Corporation - initial API and implementation
 #*******************************************************************************
--include= ~../cnf/resources/bnd/transform.props
+-include= ~../cnf/resources/bnd/transform10.props
 
-Bundle-Name: WAS JSP 2.3 Jakarta
-Bundle-SymbolicName: com.ibm.ws.jsp.2.3.jakarta
+Bundle-Name: WAS JSP 3.1 Jakarta
+Bundle-SymbolicName: com.ibm.ws.jsp.3.1.jakarta
 Bundle-Description: WAS JSP 2.3, version ${bVersion}; Jakarta Enabled
```
Comment (tevans78) [inline dev/com.ibm.ws.jsp.2.3/transformed10.bnd]:
```suggestion
Bundle-Description: WAS JSP 3.1, version ${bVersion}; Jakarta Enabled
```

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

File: dev/io.openliberty.jakarta.servlet.jsp.jstl.3.0.internal/.project [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>io.openliberty.jakarta.servlet.6.0</name>
```
Comment (volosied) [inline dev/io.openliberty.jakarta.servlet.jsp.jstl.3.0.internal/.project]: Correct this

Comment (volosied) [reply dev/io.openliberty.jakarta.servlet.jsp.jstl.3.0.internal/.project]: Rename project to use tags 3.0

File: dev/io.openliberty.org.apache.taglibs.standard.3.0/.project [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>io.openliberty.jakarta.servlet.6.0</name>
```
Comment (volosied) [inline dev/io.openliberty.org.apache.taglibs.standard.3.0/.project]: Correct this too

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

File: dev/io.openliberty.jstl.2.0.facade/bnd.bnd [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#*******************************************************************************
+<<<<<<< HEAD
+# Copyright (c) 2021, 2022 IBM Corporation and others.
```
Comment (pnicolucci) [inline dev/io.openliberty.jstl.2.0.facade/bnd.bnd]: This is a bad merge.

File: dev/io.openliberty.jstl.2.0.facade/ee10.bnd [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  org.apache.taglibs.standard.tag.el.fmt;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.tag.el.sql;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.tag.el.xml;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.tag.rt.core;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.tag.rt.fmt;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.tag.rt.sql;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.tag.rt.xml;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.tei;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.tlv;version="[3.0,4.0)", \
+  org.apache.taglibs.standard;version="[3.0,4.0)"
```
Comment (pnicolucci) [inline dev/io.openliberty.jstl.2.0.facade/ee10.bnd]: new line

File: dev/io.openliberty.org.apache.taglibs.standard.3.0/bnd.bnd [hunk 7]
```diff
@@ -0,0 +1,41 @@
+#*******************************************************************************
+# Copyright (c) 2021 IBM Corporation and others.
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.taglibs.standard.3.0/bnd.bnd]: New bnd should be a 2022 copyright date.

File: dev/io.openliberty.org.apache.taglibs.standard.3.0/build.gradle [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+task deleteJars(type:Delete) {
+    doLast {
+        println "deleteJars task, deleting " + downloadSrcDir
+        file(downloadSrcDir).deleteDir()
+    }
+}
+
+clean {
+    dependsOn deleteJars
+}
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.taglibs.standard.3.0/build.gradle]: new line

File: dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+import java.text.DateFormat;
+import java.util.Locale;
+
+/**
+ *  Smoke Tests for JSTL EE10 Tags
+ */
+@RunWith(FATRunner.class)
+public class TagsTestsEE10 {
+    private static final String APP_NAME = "TestTagsEE10";
+    // private static final String IMPORT_APP_NAME = "TestJSTLImport";
```
Comment (pnicolucci) [inline dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java]: delete?

File: dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        LOG.info("Servlet response : " + response.getText());
+
+        assertTrue("Something went wrong with the xml tags", response.getText().contains("IBM is trading above 100"));
+
+    }
+
+    /**
+     * Tests Import of SQL Tag Library
+     * 
+     * TODO: Add DB to test tags?
```
Comment (pnicolucci) [inline dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java:140]: Do we have an issue opened?

File: dev/io.openliberty.pages.3.1.internal_fat/publish/servers/tagsServer/server.xml [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+    <include location="../fatTestPorts.xml"/>
+
+    <featureManager>
+        <feature>pages-3.1</feature>
+    </featureManager>
+
+    <javaPermission className="java.util.PropertyPermission" name="dtm.debug" actions="read"/>
+
+     <javaPermission className="java.util.PropertyPermission" name="user.dir" actions="read"/>
```
Comment (pnicolucci) [inline dev/io.openliberty.pages.3.1.internal_fat/publish/servers/tagsServer/server.xml]: Fix indentation please.

File: dev/io.openliberty.pages.3.1.internal_fat/test-applications/TestTagsEE10.war/resources/WEB-INF/web.xml [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    http://www.eclipse.org/legal/epl-v10.html
+   
+    Contributors:
+        IBM Corporation - initial API and implementation
+ -->
+<web-app
+    version="3.1"
+    xmlns="http://xmlns.jcp.org/xml/ns/javaee"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd">
```
Comment (pnicolucci) [inline dev/io.openliberty.pages.3.1.internal_fat/test-applications/TestTagsEE10.war/resources/WEB-INF/web.xml]: Why are we a 3.1 version webapp and not 6.0?

File: dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+import componenttest.custom.junit.runner.Mode;
+import componenttest.custom.junit.runner.Mode.TestMode;
+
+// For testFormatTags
+import java.util.Date;
+import java.text.DateFormat;
+import java.util.Locale;
+
+/**
+ *  Smoke Tests for JSTL EE10 Tags
```
Comment (pnicolucci) [inline dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java:44]: Are these just the EE9 tests in the EE10 bucket?

https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSTLTests.java

If we repeat the JSP 2.3 bucket for EE10 why are these necessary? Are they different in some way?

Comment (volosied) [reply dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java:44]: This EE10 class test the new jakarta.tags.* URIs.  The EE9 one was for the older sun URIs. 

I think the older URIs are also tested again EE10 in the com.ibm.ws.jsp2.3_fat (JSTLTests.java) because we kept backwards compatibility.

Comment (volosied) [reply dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java:44]: But otherwise, the tests are exactly the same.

File: dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java (hunk 10 above)
Comment (volosied) [reply dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java:140]: Created: https://github.com/OpenLiberty/open-liberty/issues/21719

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

File: dev/io.openliberty.jstl.2.0.facade/ee10.bnd [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+  jakarta.servlet.jsp.tagext;version="[3.1,4.0)", \
+  jakarta.servlet.jsp.jstl.core;version="[3.0,4.0)", \
+  jakarta.servlet.jsp.jstl.fmt;version="[3.0,4.0)", \
+  jakarta.servlet.jsp.jstl.sql;version="[3.0,4.0)", \
+  jakarta.servlet.jsp.jstl.tlv;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.functions;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.lang.jstl;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.lang.jstl.parser;version="[3.0,4.0)", \
+  org.apache.taglibs.standard.lang.jstl.test;version="[3.0,4.0)";resolution:=optional, \
+  org.apache.taglibs.standard.lang.jstl.test.beans;version="[3.0,4.0)";resolution:=optional, \
```
Comment (volosied) [inline dev/io.openliberty.jstl.2.0.facade/ee10.bnd]: These should two imports should actually be removed..

**Comment by volosied** [comment]
Abandoned last build and restarted a new one: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_-CUUAAF-Ee2_YeU2rrO1jw

File: dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java [hunk 15]
```diff
@@ -260,6 +261,7 @@ public void testEL30PropertyNotWritableException() throws Exception {
      */
     @Test
     @ExpectedFFDC("javax.el.PropertyNotWritableException")
+    @SkipForRepeat(SkipForRepeat.EE10_FEATURES)
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java:264]: In the PR comments you say these are skipped due to changes in the Pages API , can you point me to the issue where we'll investigate these and work to reenable / update the tests for the Pages 3.1 feature?

File: dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/FATSuite.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.ibm.ws.fat.util.FatLogHandler;
+import io.openliberty.pages31.fat.tests.TagsTestsEE10;
+
+import componenttest.rules.repeater.EmptyAction;
+import componenttest.rules.repeater.FeatureReplacementAction;
+import componenttest.rules.repeater.RepeatTests;
+
+/**
+ * Pages 3.0 Tests
```
Comment (pnicolucci) [inline dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/FATSuite.java]: 3.1

File: dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java (hunk 13 above)
Comment (pnicolucci) [reply dev/io.openliberty.pages.3.1.internal_fat/fat/src/io/openliberty/pages31/fat/tests/TagsTestsEE10.java:44]: Can you add a comment with that information so when someone looks at this test class it is obvious to them?

File: dev/io.openliberty.pages.3.1.internal_fat/test-applications/TestTagsEE10.war/resources/META-INF/permissions.xml [hunk 17]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (pnicolucci) [inline dev/io.openliberty.pages.3.1.internal_fat/test-applications/TestTagsEE10.war/resources/META-INF/permissions.xml:1]: copyright?

File: dev/wlp-jakartaee-transform/rules/jakarta-renames.properties [hunk 18]
```diff
@@ -67,7 +67,6 @@ com.sun.xml.internal.messaging.saaj.util.transform=com.sun.xml.messaging.saaj.ut
 io.grpc=io.grpc
 io.grpc.netty=io.grpc.netty
 io.openliberty.cdi.spi=io.openliberty.cdi.spi
-jakarta.el=jakarta.el
```
Comment (pnicolucci) [inline dev/wlp-jakartaee-transform/rules/jakarta-renames.properties:70]: Can you just explain to me why this is not necessary any longer?

File: dev/com.ibm.websphere.javaee.jsp.tld.2.2/bnd.bnd [hunk 19]
```diff
@@ -26,7 +26,7 @@ Export-Package: \
     org.apache.taglibs.standard.tag.rt.sql;version="1.0.0", \
     org.apache.taglibs.standard.tag.rt.xml;version="1.0.0"
 
--includeresource: @${repo;com.ibm.ws.javax.j2ee:jsp.tld.standard;1.2}
+-includeresource: @${repo;com.ibm.ws.javax.j2ee:jsp.tld.standard;1.2;EXACT}
```
Comment (pnicolucci) [inline dev/com.ibm.websphere.javaee.jsp.tld.2.2/bnd.bnd:29]: Can you explain the addition of EXACT?

File: dev/com.ibm.ws.jsp.2.3_fat/bnd.bnd [hunk 20]
```diff
@@ -27,7 +27,12 @@ tested.features:\
     expressionLanguage-4.0,\
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsp.2.3_fat/bnd.bnd:27]: Copyright update?

File: dev/io.openliberty.jstl.2.0.facade/original.bnd [hunk 21]
```diff
@@ -49,4 +46,4 @@ Import-Package: \
   org.apache.taglibs.standard.tag.rt.xml;version="[2.0,3.0)", \
   org.apache.taglibs.standard.tei;version="[2.0,3.0)", \
   org.apache.taglibs.standard.tlv;version="[2.0,3.0)", \
-  org.apache.taglibs.standard;version="[2.0,3.0)"
+  org.apache.taglibs.standard;version="[2.0,3.0)"
```
Comment (pnicolucci) [inline dev/io.openliberty.jstl.2.0.facade/original.bnd]: newline

File: dev/io.openliberty.jakarta.servlet.jsp.jstl.3.0.internal/.project (hunk 3 above)
Comment (pnicolucci) [reply dev/io.openliberty.jakarta.servlet.jsp.jstl.3.0.internal/.project]: This looks good now!

File: dev/io.openliberty.org.apache.taglibs.standard.3.0/.project (hunk 4 above)
Comment (pnicolucci) [reply dev/io.openliberty.org.apache.taglibs.standard.3.0/.project]: Looks good!

File: dev/com.ibm.websphere.javaee.jsp.tld.2.2/bnd.bnd (hunk 19 above)
Comment (volosied) [reply dev/com.ibm.websphere.javaee.jsp.tld.2.2/bnd.bnd:29]: bnd treats the the version,  1.2  in this case, more as a minimum.  So if we uploaded version 2.0 for the same group id and archifact id, then it would pull this later version in instead. 

I ran into this problem with the jakarta tags 2.0 bundle after I uploaded the 3.0 jar.  

I just added EXACT to all the bnds for consistency.

File: dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java (hunk 15 above)
Comment (volosied) [reply dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java:264]: https://github.com/OpenLiberty/open-liberty/issues/21743

File: dev/wlp-jakartaee-transform/rules/jakarta-renames.properties (hunk 18 above)
Comment (volosied) [reply dev/wlp-jakartaee-transform/rules/jakarta-renames.properties:70]: In the EL 5.0 PR, I needed to transform the "io.openliberty.jakarta.jstl.2.0" bundle in order to get features which depended on EL to resolve. 

By transforming o.openliberty.jakarta.jstl.2.0 (an EE9 bundle),  I needed the jakarta.el imports to update. However, the current transformation rules only applied to javax.el -> jakarta.el.  By stating jakarta.el=jakarta.el, the versions then updated. 

We have a new jstl 3.0 api, so this is rule is no longer necessary.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Review by pnicolucci — APPROVED** [review]

**Comment by pnicolucci** [comment]
#libby

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat,io.openliberty.pages.3.1.internal_fat

**Comment by volosied** [comment]
Squashed all the commits into one. Merging now..

**Comment by volosied** [comment]
#libby
